Client discovery + basic light API

radar opened the pull request:
Taking inspiration from @bjeanes' Playground code here: http://play.golang.org/p/uFj1mdBKG6.

What this is attempting to do is build a binary representation from a message (i.e. the opposite of the `Decode` function), in preparation for sending it back down the pipe and out to the world.

I still have no idea what I'm doing here, but this _feels_ like the right track.

radar committed:
WIP: MarshalBinary

bjeanes commented on protocol/message.go:
```diff
@@ -56,8 +57,20 @@ func (msg *message) UnmarshalBinary(data []byte) error {
-	return
+
+	raw := m.Header.ToRawHeader()
+	size := unsafe.Sizeof(m.Header) + unsafe.Sizeof(m.Payload)
```
We'd need `unsafe.Sizeof(raw)` here

bjeanes replied:
Also, unless Go specifically works around this, `unsafe.Sizeof(m.Header)` might be returning the size of the pointer itself which would be a constant size, anyway.

radar replied:
The header is indeed a constant size. It's the payload which is variable. Since the `raw` is just the header, calling `unsafe.Sizeof(raw)` would not return the correct size for the complete message, which is the size of the header + payload data on the end. This is why I've added together both the size of the header and the payload.

Do you still think this is wrong?

bjeanes replied:
Yes. You are taking the size of `m.Header` which is the expanded header, when it needs to be of `raw`. I didn't mean to imply replacing the payload size. It's definitely header+payload, but you're taking the size of the wrong header.

My comment regarding constant size was about the pointer. If `Sizeof` works the way I remember C working, it'd take the size of the pointer, which will just be a byte or two every time, no matter what it's pointing to (**EDIT**: See [this playground](http://play.golang.org/p/IEHyQTglUE)). It's irrelevant anyway, since we don't want the size of that type of header.

So summary, this line should be:

``` go
size := unsafe.Sizeof(raw) + unsafe.Sizeof(m.Payload)

// or

size := SOME_CONSTANT_WE_SET + unsafe.Sizeof(m.Payload)
```

radar replied:
Ah, I see. I'll change this over now.

bjeanes replied:
To avoid using `unsafe`, we could also just dump out the payload into a different buffer, check its size, update the header, write the header out, then concatenate the two byte arrays, but doesn't matter so much I think.

bjeanes commented:
This looks basically right to me, other than the missing fields obviously.

radar commented on protocol/message.go:
```diff
@@ -56,8 +57,21 @@ func (msg *message) UnmarshalBinary(data []byte) error {
+
+	raw := m.Header.ToRawHeader()
+	// size := unsafe.Sizeof(raw) + unsafe.Sizeof(m.Payload)
+	raw.Size = 36 // TODO: I can't get the above to give me the numbers I want, so I'm hardcoding (for now).
```
I tried using `unsafe.Sizeof` as we discussed, but as the comment says I can't get it to give me the right numbers.

bjeanes replied:
I played a bit with it too. It seems to always give 40 for structs and I think it's because there are a few caveats:
- it is returning the size in memory, not the size it would be upon encoding. So it might be encoding some extra type information or pointer addresses, who knows... Go could also be aligning struct fields on memory word boundaries so padding some short fields out to a longer in memory size. I don't know if Go is doing this, but C will sometimes.
- it doesn't encode the size of data in slices, just the slice "header" information (length, capacity, type, etc)

Furthermore `m.Payload` is essentially an `interface{}` type, which behind the scenes is a couple of pointers. It looks like `unsafe.Sizeof` is always returning `16` there, though there is probably a way to get at the root value (`reflect` package maybe?).

So, instead I think we should do something like this:

``` go
// probably in header.go
const HEADER_SIZE = 36

// some error handling etc ommitted
func (m *message) MarshalBinary() (data []byte, err error) {
    buf := bytes.NewBuffer([]byte{})
    binary.Write(buf, binary.LittleEndian, m.Payload)
    payloadBytes := buf.Bytes()

    raw := m.Header.ToRawHeader()
    raw.Size = HEADER_SIZE + len(payloadBytes)
    raw.Type = m.Payload.Id()
    buf = bytes.NewBuffer([]byte{}) // new buffer
    binary.Write(buf, binary.LittleEndian, raw)

    return append(buf.Bytes(), payloadBytes...), nil // obv actually with the error handling heh
}
```

Totally untested and just typed up in this comment, but something like that might work. Assuming `binary.Write` doesn't give a headache with writing out the interface type.

bjeanes replied:
This [`.patch`](http://take.ms/3FItS) file applied to your HEAD works, I think, but for posterity here's the main chunk:

``` go
func (m *message) MarshalBinary() ([]byte, error) {
    buf := bytes.NewBuffer([]byte{})
    err := binary.Write(buf, binary.LittleEndian, m.Payload)
    if err != nil {
        return []byte{}, err
    }
    payloadBytes := buf.Bytes()

    raw := m.Header.ToRawHeader()
    raw.Size = HEADER_SIZE + uint16(len(payloadBytes))
    raw.Type = m.Payload.Id()
    buf.Truncate(0)
    err = binary.Write(buf, binary.LittleEndian, raw)

    if err != nil {
        return []byte{}, err
    }

    return append(buf.Bytes(), payloadBytes...), nil
}
```

radar commented:
All the fields are there now. I'm just unsure about the Size thing.

bjeanes commented on protocol/message_test.go:
```diff
@@ -392,3 +395,54 @@ func TestDecodeDeviceStateTime(t *T) {
+
+	// t.Log("\n" + hex.Dump(data))
+	// It should be this:
+	// DATA: length=36
```
Actually this line wouldn't print out with `hex.Dump(data)` :P

radar replied:
Pedant! :D

bjeanes replied:
Dude... I am such a pedant, for sure :P. I think it pays off though, even if it's annoying. Hopefully ...

bjeanes commented on protocol/message.go:
```diff
@@ -56,8 +57,21 @@ func (msg *message) UnmarshalBinary(data []byte) error {
+	raw.Type = m.Payload.Id()
+
+	buf := bytes.NewBuffer([]byte{})
+	err = binary.Write(buf, binary.LittleEndian, raw)
```
You're not writing out the `Payload` yet. This particular payload is 0-size so the test passes, but it will create invalid output for other types of payloads.

radar committed:
Implement connection.Write

Clients should use this method to write out any data.

radar committed:
Export Message type and fields

This is so that it's possible to build messages in the client's code

radar committed:
Implement client.Discover()

radar commented:
I've made some more changes to this today, mainly implementing `connection.Write()` and some of the client code so you can see the direction I'm taking on this. It might be _slightly_ tangential to this PR, but imo it's pretty closely related.

What I'm thinking of doing is moving `c.connection.Write()` to its own function on the client, as well as the bulk of the message building. That's for later.

For now, this code coupled with [this Gist](https://gist.github.com/radar/c3ed63ef70e10d1d5cdd) sends out a correctly formatted `DeviceGetPanGateway` message to the network, and I can see my bulb is responding to that.

bjeanes commented:
:boom: Nice. I am not yes sure I visualize what you mean by moving the connection stuff to the client, but I guess I will when you have an example. As long as the protocol package can still encode/decode stuff without the higher level client. Maybe it's an uneccesary goal, but I'd like to have the protocol implementation be able to stand-alone and the client just pull it altogether in a useable API.

radar commented:
I mean like ^ that commit. Just moving it into a non-exported function means that we can send it along a payload and the function should do the rest. I am not sure if we have to do stuff regarding site/target yet, but I guess we can figure that out later.

> Maybe it's an uneccesary goal, but I'd like to have the protocol implementation be able to stand-alone and the client just pull it altogether in a useable API.

How would we go about that?

bjeanes commented:
That commit looks like a good direction to experiment!

> How would we go about that?

Basically, eventually, the client shouldn't know anything about UDP connections or specific bytes on the wire. `protocol` handles turning Go structs into bytes and transmitting them and vice versa, and that's about it. `client` would handle creating/understanding those structs in terms of higher level concepts like turning lights on/off and changing colors etc and shouldn't know anything about or directly expose the bytes or networkiness. `cli` should just use that library to provide a bridge for humans to do those actions without coding :)

radar commented:
Using the code as it is now, I've gotten it to a point where the API to turn a light on / off is in place. One problem: it doesn't bloody work. I'm probably sending through the wrong parameters _somewhere_. It would be beneficial for me if `lifx-snoop` snoop'd on outgoing packets as well as packets from the bulb, so that's where I'm probably going to look into next.

radar committed:
Correct bitshifting of tagged field

radar committed:
create three sockets: peer, broadcast and write

peer + broadcast are for reading those messages respectively, and write is for writing out new messages. ListenMulticastUDP does not seem to support Write operations... it just does nothing when Write is called on it.

radar committed:
Pass only site + addressable(true) + tagged(true) in connection.Write

This makes my client code work... but I think what we'll need to do is flag the message as addressable someway. I reckon what we'll need is struct for client message building which is then read and converted into an internal struct

radar committed:
Add tests for tagged MarshalBinary + SetPowerOnMarshalBinary

radar commented on protocol/header.go:
```diff
@@ -41,17 +41,33 @@ func (h *header) version() uint16 { return 0xfff & uint16(h.Bitfield1) }
+	raw.AtTime = h.AtTime
+
+	addressable := btou(h.Addressable) << 12
+	tagged := btou(h.Tagged) << 13
```
@bjeanes: I changed this value from 14 to 13 because it was not sending through the correct information otherwise. Was there any particular reason you used 14?

bjeanes replied:
Hmm it may have been a typo or I may have been screwing around in a console and using signed ints instead of unsigned ints which would have thrown off the value I was targeting (`0x4000`).

radar replied:
Well, with 14 it doesn't work, and with 13 it does. So I'm going to leave it as 13 unless there's a good reason.

bjeanes replied:
:+1:  There's no good reason. I either made a mistake via a typo or via using a signed int instead of an unsigned int, wherein 14 probably would work but we aren't using signed here. Sorry if that wasn't clear.

radar commented:
I think the only two things left to do on this are:
- Only capture unique bulbs based on target info
- `go fmt`

Can you think of anything else to do?

bjeanes commented:
Not really. The functionality looks great. I can't wait to get this in.

As per my totally pedantic style, I'd want to push this commits around a bit such that there were as few as possible (ideally, none) WIP/broken commits so that each one was a an accretion in working functionality and git-bisect remains our friend, etc.

radar committed:
`go fmt` for message_test.go

radar commented:
Squashed down into 10 individual commits which implement the various parts of these changes. I'll see if I can get another lamp from Ikea or somewhere to test the targeting out properly.

radar committed:
Read from both broadcast+peer sockets

Mainly for the benefit of lifx-snoop, but perhaps clients care about messages on both these sockets too

bjeanes commented:
OK I did a bit more rebasing on here to get rid of the WIP commits and partial implementations. So that I don't fuck anything up or in case you disagree with my re-arrangements, I've pushed to a different branch and opened PR #9. If that looks good to you, feel free to hit merge.

Here's what I did:
- Moved "Export Message type and fields" commit to the very top 
- Run gofmt on each commit individually to get rid of "`go fmt` for message_test.go" commit
- Squashed "Correct bitshifting of tagged field" into "WIP: MarshalBinary" (it was just fixing something in the latter)
- Squashed "Implement client.Discover()" and "Implement connection.Write" (the former was the first place it would have been used but replaced it with `connection.WriteMessage` anyway)
- Renamed "WIP: MarshalBinary" to "Implement MarshalBinary for Message"
- Used [`run-command-on-git-revisions`](https://github.com/garybernhardt/dotfiles/blob/master/bin/run-command-on-git-revisions) to make sure tests passed at every commit (`run-command-on-git-revisions master HEAD 'go test ./...'`)

bjeanes commented:
I'm gonna go ahead and merge the other version in anyway, actually. Authorship etc is preserved and if I did miss anything, that can always be added!
